tests/zep/test_metadatasource.py

```diff
@@ -0,0 +1,124 @@
+    # setup
+    sources = get_context().sources
+    yield sources
+    _SourceFactories._SourceFactories__source_factories.clear()
+    _SourceFactories.type_lookup.clear()
+
+
+class TestMetaDatasource:
+    def test__new__registers_sources_factory_method(
+        self, context_sources_clean: _SourceFactories
+    ):
+        expected_method_name = "add_my_test"
+
+        ds_factory_method_initial = getattr(
+            context_sources_clean, expected_method_name, None
+        )
+        assert ds_factory_method_initial is None, "Check test cleanup"
+
+        class MyTestDatasource(metaclass=MetaDatasource):
+            pass
+
+        ds_factory_method_final = getattr(
+            context_sources_clean, expected_method_name, None
+        )
+
+        assert (
+            ds_factory_method_final
+        ), f"{MetaDatasource.__name__}.__new__ failed to add `{expected_method_name}()` method"
+
+    def test__new__injects_datasource_add_asset_method(
+        self, context_sources_clean: _SourceFactories
+    ):
+        expected_add_asset_method_name = "add_magic_asset"
+
+        class MagicAsset(DataAsset):
+            pass
+
+        class MyTestDatasource(metaclass=MetaDatasource):
+            asset_types = [MagicAsset]
+
+        my_datasource = MyTestDatasource()
+
+        asset_method = getattr(my_datasource, expected_add_asset_method_name, None)
+        assert asset_method, (
+            f"{MetaDatasource.__name__}.__new__"
+            " failed to inject `{expected_add_asset_method_name}()` method to the Datasource class"
+        )
+
+    def test__new__updates_asset_type_lookup(
+        self, context_sources_clean: _SourceFactories
+    ):
+        type_lookup = context_sources_clean.type_lookup
+
+        class FooAsset(DataAsset):
+            pass
+
+        class BarAsset(DataAsset):
+            pass
+
+        class FooBarDatasource(metaclass=MetaDatasource):
+            asset_types = [FooAsset, BarAsset]
+
+        print(f" type_lookup ->\n{pf(type_lookup)}\n")
+
+        asset_types = [FooAsset, BarAsset]
+
+        registered_type_names = [type_lookup.get(t) for t in asset_types]
+        for type_, name in zip(asset_types, registered_type_names):
+            print(f"`{type_.__name__}` registered as '{name}'")
+            assert name, f"{type.__name__} could not be retrieved"
+
+        assert len(asset_types) == len(registered_type_names)
+
+
+def test_minimal_ds_to_asset_flow(context_sources_clean):
+    # 1. Define Datasource & Assets
+    class BlueAsset(DataAsset):
+        pass
+
+    class RedDatasource(metaclass=MetaDatasource):
+        assets: MutableMapping[str, DataAsset]
+        asset_types = [BlueAsset]
+
+        def __init__(self, name: str) -> None:
+            self.name = name
+            self.assets = {}
+
+        def get_asset(self, asset_name: str) -> DataAsset:
+            return self.assets[asset_name]
```

Comment:
It would be nice if these methods could be inherited.

Normally we could actually define a default implementation for the `get_asset()` method, and the implementor could choose to use it or use their own.

```python
    class BlueAsset(DataAsset):
        pass

    class RedDatasource(Datasource, metaclass=MetaDatasource):
        assets: MutableMapping[str, DataAsset]
        asset_types = [BlueAsset]

```

But doing so we get this error `TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases`

We could deal with this in a couple of ways...
1. Make `Datasource` an ABC instead of `Protocol`
2. Inject the `get_asset` method as part of `MetaDatasource.__new__`
3. ...

Reply:
This is a really good point. I think we should change Datasource to not be a protocol now that we need implementation. I haven't tried this but reading the docs maybe something like this would work:

Make `MetaDatasource` a subclass of ABC. Make a new `Datasource` class whose metaclass is `MetaDatasource`. Implementers now inherit from `Datasource` and implement the abstract methods there. [A quick google shows one thing to do when subclassing](https://stackoverflow.com/questions/43858592/how-to-subclass-abcmeta) `ABC` is to make the `new` and `__init__` signatures align.
```
class MetaDatasource(metaclass=ABC):
   ...

class Datasource(metaclass=MetaDatasource)
  # This is now an abstract class and also has our MetaDatasource logic

# Datasource implementor does this
class MyDatasource(Datasource):
  ...
```

Reply:
Following up on the discussion yesterday.
This is the method we are going to pursue, but we will need to add a conditional to `MetaDatasource.__new__` to check if the `cls` is a concrete `Datasource` or the `DatasourceABC`.
```python
if CLASS is DATASOURCE_ABC:
   return super().__new__()
else:
    DO_FANCY_HOOK_STUFF
```